### ISSUE TITLE
l3doc: variable not indexed correctly

### ISSUE THREAD

**Issue by stone-zeng** [latex3/latex3#450]
I have a variable named `\g_foo_seq`, which should be indexed in the "foo commands" group. But now it is incorrectly indexed in the "seq commands" group. However, something like `\c_foo_bar_seq` is indexed correctly, i.e. in "foo commands".

**Comment by josephwright** [comment]
That's I think by-design: names with only one part are used by the kernel, _e.g._ `\l_tmpa_tl`.

**Comment by Stone-Zeng** [comment]
Maybe those only with specific names like `tmpa`, `tmpb`, `zero`, `one` etc should be considered as kernel variables.

**Comment by josephwright** [comment]
@Stone-Zeng  Quite possibly, and certainly we should mention it :) @wspr is currently busy with 'real life', so we might have to wait a little while for an official position (he looks after `l3doc`).

**Comment by blefloch** [comment]
Possibly tmpa, tmpb, empty, zero should also be in l3prefixes.csv

**Comment by josephwright** [comment]
Probably true! Also reminds me we need to sort the integer constants ...

**Comment by blefloch** [comment]
I think this is the full list of fake modules that we may want to
reserve for l3kernel: alignment, ampersand, atsign, backslash, catcode,
circumflex, colon, dollar, e, empty, hash, inf, left, log, math, mark,
max, minus, nan, nil, no, novalue, parameter, percent, pi, recursion,
right, space, stop, term, tilde, tmpa, tmpb, underscore, zero, one, two,
three, four, five, six, seven, eight, nine, ten, eleven, twelve,
thirteen, fourteen, fifteen, sixteen, thirty

**Comment by josephwright** [comment]
@blefloch  Well some of those likely need to be dropped (we've discussed the various integer constants ...)

**Comment by blefloch** [comment]
The problem with a fixed set of exceptions is that we have quite many, for instance things like `\c_ampersand_str`, `\c_underscore_str` etc.

What about adding a key `module` to the `function` and `macro` environments, that would override the module? (I think we already have that for `\cs{...}` so the framework is there.)  Then drop completely the special case for such short variable names, so that the code would by default put `\c_empty_tl` into module `empty`, but the key `module = tl` would override that.

Before I implement I'd like others' thoughts.

**Comment by stone-zeng** [comment]
Now I have found some issues with the documentation of `l3doc` itself (compiled by myself since it's not provided):

- The `@@` in this code fragment (and the following) are not correct (you can move the comments outside):
https://github.com/latex3/latex3/blob/2457ff211d37178076e6b1f48a19d1ea2cc9c9b1/l3kernel/l3doc.dtx#L3826-L3841
![image](https://user-images.githubusercontent.com/18233088/61306901-ed49cc80-a81f-11e9-9179-76149a4beb5e.png)
- There are some incorrect index terms such as
  - `\pkg{#2}` internal commands, from the above fragment
  - `\pkg{@tempa}` commands, but it should be under *TeX and LaTeX2ε commands* (I guess it's due to the `:=`):
 https://github.com/latex3/latex3/blob/2457ff211d37178076e6b1f48a19d1ea2cc9c9b1/l3kernel/l3doc.dtx#L3690-L3693
  - `\pkg{\meta {name}}` commands:
https://github.com/latex3/latex3/blob/2457ff211d37178076e6b1f48a19d1ea2cc9c9b1/l3kernel/l3doc.dtx#L267-L269
which will lead to error when compiled.
- Other small issues in the index:
  - `\>` becomes `\!\efill`
  - The escape chars (e.g. `\A`, `\d`) in regex should not be indexed as control sequences:
https://github.com/latex3/latex3/blob/2457ff211d37178076e6b1f48a19d1ea2cc9c9b1/l3kernel/l3doc.dtx#L1992-L1995
https://github.com/latex3/latex3/blob/2457ff211d37178076e6b1f48a19d1ea2cc9c9b1/l3kernel/l3doc.dtx#L2436-L2437
  - Only `\\` rather than `\\xxx` are indexed in `l3doc.ist` (actually I think they shouldn't be indexed):
https://github.com/latex3/latex3/blob/2457ff211d37178076e6b1f48a19d1ea2cc9c9b1/l3kernel/l3doc.dtx#L4529-L4545

Really too complicated…

**Comment by PhelypeOleinik** [comment]
@stone-zeng I made a pull request (#602) addressing some of the issues you pointed out.

The `<@@=...>` issue was introduced with #596 (my fault :-). Certainly, the comments can be moved outside the code block, but one can also use `@@@@`, which in the generated code/documentation becomes `@@`. Note that this bug wouldn't affect the actual code in the same way it would the documentation... Perhaps we should further change the behaviour of `<@@=...>` to switch the module name only when preceded by a new line and a `%` to mimick exactly the behaviour of `l3docstrip`;

> `\pkg{#2}` internal commands, from the above fragment

Sorry, I didn't understand what you meant with this comment.

Yes, `\@tempa:` is a single control sequence there. I replaced the entire `\@for` with a more expl3-y `\clist_map_inline:Nn`.

I added the `no-index` option to `\cs{\meta{name}:\meta{signature}}` to prevent indexing.

> `\>` becomes `\!\efill`

Additionally `=` and `|` would fail to index because `makeindex` would reject these entries. I added a quotation to fix this issue.

I'm not sure how to prevent the indexing of `\A` and `\d` in those circumstances, neither how to avoid indexing the `\\` in the `gind.ist` file... Ideas?

**Comment by stone-zeng** [comment]
- The `\pkg{#2}` issue is just the `<@@=...>`, where `@@` is replaced by `#2` and will lead to error in index. It seems that you have fixed it in #602.
- There is a "verbatim mode" in `docstrip` (see `docstrip.pdf` Sec. 6.3). Maybe we can use these mode for `gind.ist` etc., and turn off the index function in this block.

**Comment by PhelypeOleinik** [comment]
Thanks @stone-zeng, I added another commit to #602 to address the issue of indexing regex escape sequences.

The verbatim mode of `docstrip` won't work here because the indexing happens at a later stage of the processing, long after `docstrip` finished its job. The code in `l3doc.dtx` is scanned once by `docstrip` to generate `l3doc.cls`, and once again by `l3doc.cls` to parse the code (detecting internals, replacing `@@` by the module name, and indexing commands).

To make that work I added a `do-not-index` option to the `macro` environment which basically does `\DoNotIndex` to its arguments. I figured an option would be best because such removals are in a per-case basis, not something that can be automated. I also didn't add the option to the `function` environment, as it's supposed to be used to document code-level functions, so they should be indexed (of course adding there is easy, if needed).

I think this covers the pointed issues. I only need to figure out why some test are failing after these changes, other than that I think it's good to go.

However, I'd like someone else's opinion before (of even if) I push these changes to `l3doc`.
